mysys/my_largepage.c

```diff
@@ -355,158 +310,152 @@ uchar* my_large_malloc_int(size_t *size, myf my_flags)
+    my_large_malloc(), i.e. my_malloc_lock() was used so we should free it
+    with my_free_lock()
+  */
+#if defined(HAVE_MMAP) && !defined(_WIN32)
   if (munmap(ptr, size))
   {
-    /* This occurs when the original allocation fell back to conventional memory so ignore the EINVAL error */
-    if (errno != EINVAL)
+    /*
+      This occurs when the original allocation fell back to conventional
+      memory so ignore the EINVAL error.
+    */
+    if (errno == EINVAL)
     {
-      fprintf(stderr, "Warning: Failed to unmap %zu bytes, errno %d\n", size, errno);
+      my_free_lock(ptr);
+    }
+    else
+    {
+      fprintf(stderr,
+              "Warning: Failed to unmap location %p, %zu bytes, errno %d\n",
+              ptr, size, errno);
     }
-    DBUG_RETURN(0);
   }
-  DBUG_RETURN(1);
-}
-#endif /* HAVE_MMAP */
-
-#if defined(HAVE_MMAP) && !defined(__linux__) && !defined(MAP_ALIGNED) \
-    && !defined(_WIN32)
-
-/* Solaris for example has only MAP_ANON, FreeBSD has MAP_ANONYMOUS and
-MAP_ANON but MAP_ANONYMOUS is marked "for compatibility" */
-#if defined(MAP_ANONYMOUS)
-#define OS_MAP_ANON     MAP_ANONYMOUS
-#elif defined(MAP_ANON)
-#define OS_MAP_ANON     MAP_ANON
-#else
-#error unsupported mmap - no MAP_ANON{YMOUS}
-#endif
-
-
-/* mmap(non-Linux,non-FreeBSD) pages allocator  */
-
-uchar* my_large_malloc_int(size_t *size, myf my_flags)
-{
-  uchar* ptr;
-  int mapflag;
-  DBUG_ENTER("my_large_malloc_int");
-
-  mapflag= MAP_PRIVATE | OS_MAP_ANON;
-
-  if (my_use_large_pages && my_large_page_size)
+#elif defined(_WIN32)
+  /*
+     When RELEASE memory, the size parameter must be 0.
+     Do not use MEM_RELEASE with MEM_DECOMMIT.
+  */
+  if (ptr && !VirtualFree(ptr, 0, MEM_RELEASE))
   {
-    /* Align block size to my_large_page_size */
-    *size= MY_ALIGN(*size, (size_t) my_large_page_size);
+    fprintf(stderr,
+            "Error: VirtualFree(%p, %zu) failed; Windows error %lu\n", ptr, size, GetLastError());
+    my_free_lock(ptr);
   }
-  ptr= mmap(NULL, *size, PROT_READ | PROT_WRITE, mapflag, -1, 0);
-  if (ptr == (void*) -1)
+#else
+  my_free_lock(ptr);
+  if (1)
   {
-    ptr= NULL;
-    if (my_flags & MY_WME)
-    {
-      fprintf(stderr,
-              "Warning: Failed to allocate %zu bytes from memory."
-              " errno %d\n", *size, errno);
-    }
   }
-  DBUG_RETURN(ptr);
-}
-#endif /* defined(HAVE_MMAP) && !defined(__linux__) && !defined(_WIN32) */
+#endif
+  /*
+    For ASAN, we need to explicitly unpoison this memory region because the OS
+    may reuse that memory for some TLS or stack variable. It will remain
+    poisoned if it was explicitly poisioned before release. If this happens,
+    we'll have hard to debug false positives like in MDEV-21239.
+    For valgrind, we mark it as UNDEFINED rather than NOACCESS because of the
+    implict reuse possiblility.
+  */
+  else
+    MEM_UNDEFINED(ptr, size);
 
-#ifdef _WIN32
+  DBUG_VOID_RETURN;
+}
```

Comment:
Wouldn't it be easier to read if it were looking like:
```
#ifdef _WIN32
  if (...) { ... DBUG_RETURN(); }
#elif HAVE_MMAP
  if (...) { ... DBUG_RETURN(); }
#endif
  my_free_lock(ptr);
  MEM_UNDEFINED(ptr, size);
```

I also wonder shouldn't `MEM_UNDEFINED()` be done by `my_free_lock()`?  And shouldn't it be `TRASH_FREE()` instead?